x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java

```diff
@@ -124,104 +143,156 @@ ResolvedIndices resolveIndicesAndAliases(IndicesRequest indicesRequest, Metadata
+                                    indicesRequest.includeDataStreams())) {
+                                resolvedIndicesBuilder.addLocal(authorizedIndex);
+                            }
                         }
-                    }
-                }
-                // if we cannot replace wildcards the indices list stays empty. Same if there are no authorized indices.
-                // we honour allow_no_indices like es core does.
-            } else {
-                final ResolvedIndices split;
-                if (allowsRemoteIndices(indicesRequest)) {
-                    split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
+                        resolvedIndicesStepListener.onResponse(resolvedIndicesBuilder);
+                    }, listener::onFailure));
                 } else {
-                    split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), Collections.emptyList());
-                }
-                List<String> replaced = indexAbstractionResolver.resolveIndexAbstractions(split.getLocal(), indicesOptions, metadata,
-                        authorizedIndices, replaceWildcards, indicesRequest.includeDataStreams());
-                if (indicesOptions.ignoreUnavailable()) {
-                    //out of all the explicit names (expanded from wildcards and original ones that were left untouched)
-                    //remove all the ones that the current user is not authorized for and ignore them
-                    replaced = replaced.stream().filter(authorizedIndices::contains).collect(Collectors.toList());
+                    // if we cannot replace wildcards the indices list stays empty. Same if there are no authorized indices.
+                    // we honour allow_no_indices like es core does.
+                    resolvedIndicesStepListener.onResponse(new ResolvedIndices.Builder());
                 }
-                resolvedIndicesBuilder.addLocal(replaced);
-                resolvedIndicesBuilder.addRemote(split.getRemote());
+            } else {
+                authorizedIndicesSupplier.getAsync(ActionListener.wrap(authorizedIndices -> {
+                    final ResolvedIndices split;
+                    if (allowsRemoteIndices(indicesRequest)) {
+                        split = remoteClusterResolver.splitLocalAndRemoteIndexNames(indicesRequest.indices());
+                    } else {
+                        split = new ResolvedIndices(Arrays.asList(indicesRequest.indices()), Collections.emptyList());
+                    }
+                    ResolvedIndices.Builder resolvedIndicesBuilder = new ResolvedIndices.Builder();
+                    // cannot pass the async supplier
+                    List<String> replaced = indexAbstractionResolver.resolveIndexAbstractions(split.getLocal(), indicesOptions, metadata,
+                            authorizedIndices, replaceWildcards, indicesRequest.includeDataStreams());
```

Comment:
There's an interesting point to be made here.
`indexAbstractionResolver.resolveIndexAbstractions` can be made to work without the `authorizedIndices` list expanded, BUT, because it has to iterate over the `split.getLocal()` list ON THE STACK.

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/72118/commits/c2466b8ee1a7fb7681569376a0f227b040ca9995 to make `indexAbstractionResolver.resolveIndexAbstractions` async.